src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -3262,6 +3262,293 @@ void C2_MacroAssembler::stringL_indexof_char(Register str1, Register cnt1, Regis
+    break;
+  case T_INT:
+    // do nothing
+    break;
+  default:
+    ShouldNotReachHere();
+  }
+}
+
+void C2_MacroAssembler::arrays_hashcode(Register ary1, Register cnt1, Register result,
+                                        Register index, Register coef, Register tmp, XMMRegister vnext,
+                                        XMMRegister vcoef0, XMMRegister vcoef1, XMMRegister vcoef2, XMMRegister vcoef3,
+                                        XMMRegister vresult0, XMMRegister vresult1, XMMRegister vresult2, XMMRegister vresult3,
+                                        XMMRegister vtmp0, XMMRegister vtmp1, XMMRegister vtmp2, XMMRegister vtmp3,
+                                        int mode) {
+  ShortBranchVerifier sbv(this);
+  assert(UseAVX >= 2, "AVX2 intrinsics are required");
+  assert_different_registers(ary1, cnt1, result, index, coef, tmp);
+  assert_different_registers(vnext, vcoef0, vcoef1, vcoef2, vcoef3, vresult0, vresult1, vresult2, vresult3, vtmp0, vtmp1, vtmp2, vtmp3);
+
+  Label SHORT_UNROLLED_LOOP_BEGIN, SHORT_SCALAR_LOOP_BEGIN,
+        LONG, LONG_INIT, LONG_SCALAR_LOOP_BEGIN, LONG_SCALAR_LOOP_END, LONG_VECTOR_LOOP_BEGIN, LONG_VECTOR_LOOP_END,
+        END;
+
+  // For "renaming" for readibility of the code
+  Register bound;
+  XMMRegister vcoef[] = { vcoef0, vcoef1, vcoef2, vcoef3 },
+              vresult[] = { vresult0, vresult1, vresult2, vresult3 },
+              vtmp[] = { vtmp0, vtmp1, vtmp2, vtmp3 };
+
+  const VectorizedHashCodeNode::HashMode hashMode = (VectorizedHashCodeNode::HashMode)mode;
+  const BasicType eltype = VectorizedHashCodeNode::adr_basic_type(hashMode);
+  const bool is_string_hashcode = hashMode == VectorizedHashCodeNode::LATIN1 || hashMode == VectorizedHashCodeNode::UTF16;
+  const bool is_unsigned = is_string_hashcode || hashMode == VectorizedHashCodeNode::CHAR;
+  const int elsize = arrays_hashcode_elsize(eltype);
+  // vector loops does not currently support short and signed byte arrays for now, need to figure out how to properly sign-extend when loading into vectors
+  const bool generate_vectorized_loop = is_unsigned || hashMode == VectorizedHashCodeNode::INT;
+
+  // int result = 0|1;
+  movl(result, is_string_hashcode ? 0 : 1);
+
+  // if (cnt1 == 0) {
+  cmpl(cnt1, 0);
+  jcc(Assembler::equal, END);
+
+  if (hashMode == VectorizedHashCodeNode::UTF16) {
+    shrl(cnt1, 1);
+  }
+
+  // } else if (cnt1 < 32) {
+
+  if (generate_vectorized_loop) {
+    cmpl(cnt1, 32);
+    jcc(Assembler::greaterEqual, LONG);
+  }
+
+  // int i = 0;
+  movl(index, 0);
+  // int bound = cnt1 & ~(4 - 1);
+  bound = coef;
+  movl(bound, cnt1);
+  andl(bound, ~(4-1));
+
+  // for (; i < bound; i += 4) {
+  bind(SHORT_UNROLLED_LOOP_BEGIN);
+  // i < bound;
+  cmpl(index, bound);
+  jccb(Assembler::greaterEqual, SHORT_SCALAR_LOOP_BEGIN);
+  for (int idx = 0; idx < 4; idx++) {
+    // h = (31 * h) or (h << 5 - h);
+    movl(tmp, result);
```

Comment:
If you are unrolling this, maybe break the dependency chain, `h = h * 31**4 + x[i] * 31**3 + x[i + 1] * 31**2 + x[i + 2] * 31 + x[i + 3]`

Reply:
A 256-bit vector is only 8 ints so this loop seems redundant, maybe running with the stride of 2 instead, in which case the single scalar calculation does also not need a loop.